src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.cs

```diff
@@ -258,38 +316,47 @@ private async Task ProcessProjectAsync(Project project, ImmutableArray<ISymbol>
         Document document,
         MetadataUnifyingSymbolHashSet symbols,
         Dictionary<ISymbol, PooledHashSet<string>> symbolToGlobalAliases,
+        Action<Reference> onReferenceFound,
         CancellationToken cancellationToken)
     {
-        await _progress.OnFindInDocumentStartedAsync(document, cancellationToken).ConfigureAwait(false);
-
-        try
+        // We're doing to do all of our processing of this document at once.  This will necessitate all the
+        // appropriate finders checking this document for hits.  We know that in the initial pass to determine
+        // documents, this document was already considered a strong match (e.g. we know it contains the name of
+        // the symbol being searched for).  As such, we're almost certainly going to have to do semantic checks
+        // to now see if the candidate actually matches the symbol.  This will require syntax and semantics.  So
+        // just grab those once here and hold onto them for the lifetime of this call.
+        var cache = await FindReferenceCache.GetCacheAsync(document, cancellationToken).ConfigureAwait(false);
```

Comment:
now that we have the cache, per document, we've done all the up front async work we need, and can do everything synchronously from that point on.